<launch>
  <param name="use_sim_time" value="true" /> #be sure to run --clock

  # CERBERUS - Frame Connector - Only needed if localization manager is not running
  <!-- <node pkg="tf" type="static_transform_publisher" name="map_corrected_tf" respawn="true" args="0 0 0 0 0 0 map camera_init_CORRECTED 10"/>
  <node pkg="tf" type="static_transform_publisher" name="corrected_init_tf" respawn="true" args="0 0 0 1.570796327 0 1.570796327 camera_init_CORRECTED camera_init 10 "/> -->
  # CERBERUS - Frame Connector

  <!-- Input topics remap -->
  <arg name="pointcloud_topic_name" default="/point_cloud_filter/lidar/point_cloud_filtered" /> #/lidar/point_cloud #/point_cloud_filter/lidar/point_cloud_filtered
  <arg name="imu_topic_name" default="/sensors/imu" />

  <!-- Config params -->
  <arg name="config_filename" default="$(find loam)/config/loam_cerberus_params.yaml" />

  <!-- LOAM Nodelet Manager  -->
  <node pkg="nodelet" type="nodelet" name="loam_nodelet_manager" args="manager" output="screen" clear_params="true"/> #launch-prefix="gdb -ex run --args" />

  <!-- LOAM Nodelet Manager
  <node pkg="nodelet" type="nodelet" name="laser_multiscan_registration_nodelet" args="load laser_multiscan_registration_nodelet loam_nodelet_manager" output="screen">
    <rosparam file="$(arg config_filename)" command="load"/>
    <remap from="/multi_scan_points" to="$(arg pointcloud_topic_name)" />
  </node>

  <node pkg="nodelet" type="nodelet" name="laser_odometry_nodelet" args="load laser_odometry_nodelet loam_nodelet_manager" output="screen">
    <rosparam file="$(arg config_filename)" command="load"/>
    <remap from="/imu" to="$(arg imu_topic_name)" />
  </node>

  <node pkg="nodelet" type="nodelet" name="laser_mapping_nodelet" args="load laser_mapping_nodelet loam_nodelet_manager" output="screen">
    <rosparam file="$(arg config_filename)" command="load"/>
    <param name="saved_submaps_filepath" value="$(find loam)"/>
    <remap from="/aft_mapped_to_init_CORRECTED" to="/loam/odometry" /> #For offline use with ANYmal stack
  </node>

  <node pkg="nodelet" type="nodelet" name="laser_transform_maintenance_nodelet" args="load laser_transform_maintenance_nodelet loam_nodelet_manager" output="screen">
    <rosparam file="$(arg config_filename)" command="load"/>
  </node>
  -->

  <node pkg="tf" type="static_transform_publisher" name="imu_to_lidar_tf" respawn="true" args="0.005303 0.037340 0.063319 0.0 0.0 0.0 1.0 lidar imu_sensor_frame  10"/>

  <node pkg="tf" type="static_transform_publisher" name="cam_to_imu_tf" respawn="true" args="0.045378 -0.073670 0.022527  -0.001728, 0.707550, -0.706662, -0.000042 imu_sensor_frame cam0_sensor_frame 10 "/>

  <!-- Laser Graph Integrator Nodelet -->
  <node pkg="nodelet" type="nodelet" name="laser_graph_integrator_nodelet" args="load laser_graph_integrator_nodelet loam_nodelet_manager" output="screen">
    <rosparam file="$(arg config_filename)" command="load"/>
    <remap from="odometry" to="/integrated_to_init_CORRECTED" />
    <remap from="odometry_status" to="/laser_integrated_status" />
    <remap from="pointcloud" to="$(arg pointcloud_topic_name)" />
    <remap from="absolute_constraints" to="/anymal/absolute_reference" />
    <remap from="anchor_constraints" to="/graph_client/anchor_nodes" />
    <remap from="submap_constraints" to="/graph_client/relative_nodes" />
  </node>
</launch>
